components/browser/storage-sync-places/src/main/java/mozilla/components/browser/storage/places/PlacesHistoryStorage.kt

```diff
@@ -0,0 +1,79 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.storage.places
+
+import android.content.Context
+import android.support.annotation.VisibleForTesting
+import kotlinx.coroutines.experimental.async
+import kotlinx.coroutines.experimental.CoroutineScope
+import kotlinx.coroutines.experimental.Deferred
+import kotlinx.coroutines.experimental.Dispatchers
+import kotlinx.coroutines.experimental.launch
+import mozilla.components.concept.storage.HistoryStorage
+import mozilla.components.concept.storage.PageObservation
+import mozilla.components.concept.storage.VisitType
+import org.mozilla.places.PlacesAPI
+import org.mozilla.places.VisitObservation
+
+class PlacesHistoryStorage(context: Context)
+    : TestablePlacesHistoryStorage(Conn.getInstance(context.getDatabasePath(DB_NAME).canonicalPath))
+
+/**
+ * Implementation of the [HistoryStorage] which is backed by a Rust Places lib via [PlacesConnection].
+ */
+@VisibleForTesting
+open class TestablePlacesHistoryStorage(private val places: PlacesAPI) : HistoryStorage {
+    private val scope by lazy { CoroutineScope(Dispatchers.IO) }
+
+    override suspend fun recordVisit(uri: String, visitType: VisitType) {
+        scope.launch {
+            places.noteObservation(VisitObservation(uri, visitType = visitType.toPlacesType()))
+        }.join()
+    }
+
+    override suspend fun recordObservation(uri: String, observation: PageObservation) {
+        // NB: visitType of null means "record meta information about this URL".
+        scope.launch {
+            places.noteObservation(
+                VisitObservation(
+                    url = uri,
+                    visitType = org.mozilla.places.VisitType.UPDATE_PLACE,
+                    title = observation.title
+                )
+            )
+        }.join()
+    }
+
+    override fun getVisited(uris: List<String>): Deferred<List<Boolean>> {
+        return scope.async { places.getVisited(uris) }
+    }
+
+    override fun getVisited(): Deferred<List<String>> {
+        return scope.async {
+            places.getVisitedUrlsInRange(
+                start = 0,
+                end = System.currentTimeMillis(),
+                includeRemote = true
+            )
+        }
+    }
+
+    override fun cleanup() {
+        Conn.close()
```

Comment:
nit: if we close the connection here: Do we want to cancel all jobs of this coroutine scope too? You may close a connection that is in use here.